3938 migration status

Eirik0 opened the pull request:
Closes #3938

Eirik0 commented:
This depends on #3888 and #3967. The last 3 commits are the ones relevant to this PR.

str4d commented on src/wallet/wallet.cpp:
```diff
@@ -2150,6 +2150,14 @@ bool CWallet::SetCryptedHDSeed(const uint256& seedFp, const std::vector<unsigned
+HDSeed CWallet::GetHDSeedForRPC() const {
+    HDSeed seed;
+    if (!pwalletMain->GetHDSeed(seed)) {
+        throw JSONRPCError(RPC_WALLET_ERROR, "HD seed not found");
```
Hmm, it just occurred to me that this, along with two other recent changes, blurs the boundary between `CWallet` and the RPC methods that use it. In the interest of moving towards `CWallet` being self-contained (and potentially a future swappable component), I don't think we should be using JSON-RPC code in `src/wallet/wallet.cpp`. However, given that we have already merged two changes that do so, I won't block on this here, as long as we don't add any more such code and make an issue to refactor this.

Eirik0 replied:
I agree. I created an issue for this.

daira replied:
Also this will break builds without ENABLE_WALLET, won't it?

Eirik0 replied:
That is a good question. I just tested this with `ENABLE_WALLET` set to false and everything built fine.

str4d replied:
The wallet RPCs are removed when `ENABLE_WALLET` is set to false, so it isn't affected by this change.

str4d commented on src/wallet/asyncrpcoperation_saplingmigration.cpp:
```diff
@@ -185,7 +185,7 @@ libzcash::SaplingPaymentAddress AsyncRPCOperation_saplingmigration::getMigration
         pwalletMain->HaveSaplingSpendingKey(fvk))) {
         // Sapling account 0 must be the first address returned by GenerateNewSaplingZKey
-        assert(pwalletMain->GenerateNewSaplingZKey() == toAddress);
+        assert(pwalletMain->GenerateNewSaplingZKey(false) == toAddress);
```
AFAICT this means that if the user has never generated a Sapling address, the migrated funds will disappear from their wallet (e.g. `z_gettotalbalance` will not show them), because this call previously had a side-effect of adding the default address into the wallet. If that side-effect is being removed here, at what point _will_ the default address be added to the wallet?

Eirik0 replied:
I added this to make it easier to test. I was also unsure whether or not it was correct to have this side-effect here. Looking at it again. I think it is preferable to leave this in. I will remove the commit with this refactoring.

str4d commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3929,6 +3930,100 @@ UniValue z_setmigration(const UniValue& params, bool fHelp) {
+        throw runtime_error(
+            "z_getmigrationstatus\n"
+            "Returns information about the status of the Sprout to Sapling migration.\n"
+            "In the result a transactions is defined as finalized iff it has ten confirmations.\n"
```
Exactly how many users are likely to realise that `iff` is not a typo? 🙃

daira replied:
All of the mathematicians :-p

daira replied:
But yes, change to `if and only if`.

Eirik0 replied:
I will change this to `if and only if`. Also, this should `if and only if it has *at least* ten confirmations`. "at least" is also missing from this definition in the ZIP.

Eirik0 commented:
I've removed the refactoring of `GenerateNewSaplingZKey`. (Part of the reason for doing this was to make it easier to get the default sapling address on demand when testing, but this was made moot by the following.) I've also update the test use an address provided by the `-migrationdestaddress` option.  

Making these changes, I noticed that the help message for `-migration` was not consistent with other help messages for boolean arguments, so I have added another commit to address that.

Eirik0 commented:
As a note note: The in the ZIP we use the word `confirmed` to mean a transaction with 10 confirmations. This was changed to `finalized` at the request of @ioptio and @nathan-at-least. I have updated https://github.com/zcash/zips/issues/229 to note that this change needs to happen there too.

Eirik0 commented:
I updated the rpc test to check the migration using both the parameter from the conf file and the default Sapling address. This is accomplished by essentially running the test twice. First, we run the test with one node which has the `-migrationdestaddress=...` (and `-migration`) parameter set. This node does a round of migrating at block 500. Then we run the test again with another node that does not have the parameters set (and therefore will use the default Sapling address) which migrates at block 1000.

daira commented on qa/rpc-tests/sprout_sapling_migration.py:
```diff
@@ -11,91 +11,161 @@
+        assert_equal(1, len(node.getrawmempool()), "mempool size at 499 % 500")
+        assert_equal(node.z_getbalance(sproutAddr), Decimal('0'))
+        assert_equal(node.z_getbalance(saplingAddr), Decimal('0'))
+        assert_true(node.z_getbalance(saplingAddr, 0) > Decimal('0'), "Unconfirmed sapling balance at 499")
```
`499 % 500`

daira commented on qa/rpc-tests/sprout_sapling_migration.py:
```diff
@@ -11,91 +11,161 @@
+        check_migration_status(self.nodes[0], False, SAPLING_ADDR, False, False, False, 0, 0)
+
+        # 1. Test using self.nodes[0] which has the parameter
+        print "Runing test using '-migrationdestaddress'..."
```
`Running`

daira commented on qa/rpc-tests/sprout_sapling_migration.py:
```diff
@@ -11,91 +11,161 @@
+        self.nodes[0].z_setmigration(False)
+
+        # 2. Test using self.nodes[1] which will use the default Sapling address
+        print "Runing test using default Sapling address..."
```
`Running`

daira commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3929,6 +3930,100 @@ UniValue z_setmigration(const UniValue& params, bool fHelp) {
+            "z_getmigrationstatus\n"
+            "Returns information about the status of the Sprout to Sapling migration.\n"
+            "In the result a transactions is defined as finalized iff it has ten confirmations.\n"
+            "Note: It is possible that manually created transactions invloving this wallet\n"
```
`involving`

daira commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3929,6 +3930,100 @@ UniValue z_setmigration(const UniValue& params, bool fHelp) {
+            "\nResult:\n"
+            "{\n"
+            "  \"enabled\": true|false,                    (boolean) Whether or not migration is enabled\n"
+            "  \"destination_address\": \"zaddr\",           (string) The Sapling address which will receive Sprout funds\n"
```
s/which/that/

daira commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3929,6 +3930,100 @@ UniValue z_setmigration(const UniValue& params, bool fHelp) {
+    LOCK2(cs_main, pwalletMain->cs_wallet);
+    UniValue migrationStatus(UniValue::VOBJ);
+    migrationStatus.push_back(Pair("enabled", pwalletMain->fSaplingMigrationEnabled));
+    //  The "destination_address" field MAY be omitted if the "-migrationaddress"
```
`-migrationdestaddress`

Eirik0 commented:
Rebased on to master.

Eirik0 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3929,6 +3930,100 @@ UniValue z_setmigration(const UniValue& params, bool fHelp) {
+            CAmount migrationAmount = 0;
+            for (const auto& js : tx.vjoinsplit) {
+                migrationAmount += js.vpub_new;
+            }
```
I think it makes sense to update the spec to include `vin.empty() && vout.empty()` in the definitions of migration transactions. Also I think it is correct to to use `-valueBalance` as opposed to `js.vpub_new`, so I will update that. Regarding `js.vpub_old`, it might make sense to change the definition to be `one or more Sprout JoinSplits all zero vpub_old fields` rather than `one or more Sprout JoinSplits with nonzero vpub_new field`, but would require more discussion so I am leaving it is for the time being.

Eirik0 replied:
One thing to note is that the changes regarding `vin`, `vout`, and the proposed change regarding `vpub_old` will only effect manually created migration transactions which may be considered unsupported as there is no way to create them using a standard zcashd node. For that reason I do not think it is super pressing to switch to using `vpub_old` rather than `vpub_new`, but it is still something to consider.

daira replied:
I've thought about this and continue to believe that the current definition in the ZIP is the simplest that satisfies the needed criteria.

daira replied:
In particular, what would be the rationale of enforcing that `vpub_old`s are zero, but not also enforcing that there are no transparent inputs? Do we also enforce that there are no transparent outputs? Where do we stop?

daira replied:
Agreed that it is correct to use `-valueBalance`. I think that gives a sensible result independent of the migration transaction criteria; what we're saying is "there are transactions that in total transferred X into Sapling, at least some of it coming from Sprout (and for transactions created by zcashd, exactly X will be from Sprout)".

Eirik0 replied:
> Where do we stop?

Good point. It is a bit of a slippery slope. I will change it back to not look at vin/vout.

daira commented on qa/rpc-tests/sprout_sapling_migration.py:
```diff
@@ -11,91 +11,165 @@
+        check_migration_status(self.nodes[0], False, SAPLING_ADDR, False, False, False, 0, 0)
+
+        # 1. Test using self.nodes[0] which has the parameter
+        print "Running test using '-migrationdestaddress'..."
```
Use a Python 3-compatible print call, i.e.
```
        print("Running test using '-migrationdestaddress'...")
```

daira commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3929,6 +3930,105 @@ UniValue z_setmigration(const UniValue& params, bool fHelp) {
+            "  \"unfinalized_migrated_amount\": nnn.n,     (numeric) The total amount of unfinalized " + CURRENCY_UNIT + " \n"
+            "  \"finalized_migrated_amount\": nnn.n,       (numeric) The total amount of finalized " + CURRENCY_UNIT + " \n"
+            "  \"finalized_migration_transactions\": nnn,  (numeric) The number of migration transactions involving this wallet\n"
+            "  \"time_started\": ttt,                      (numeric, optional) The block time of the first migration transaction\n"
```
add "as a Unix timestamp".

Eirik0 commented:
Addressed @daira's comments.

daira commented:
@zkbot r+

Eirik0 commented:
The failure was an actual failure caused by the refactoring in https://github.com/zcash/zcash/pull/3973/commits/52cfa9c1ee43e90c96e31a712b42926b857b8ee6. In `rpc_wallet_tests/rpc_wallet_z_exportwallet` we run an export on a wallet that does not have an HD Seed. The refactoring (which in hindsight should not have updated this method) caused this to throw an exception. The fix was to partially revert the aforementioned commit, removing the changes made in `rpcdump.cpp`

Eirik0 commented:
@zkbot try

mdr0id commented:
@zkbot r+

str4d commented:
Looks like a transient failure.

@zkbot retry
